src/app/clusters/operational-credentials-server/operational-credentials-server.cpp

```diff
@@ -360,6 +362,182 @@ bool emberAfOperationalCredentialsClusterAddOpCertCallback(chip::app::Command *
+
+    TLV::TLVWriter * writer = nullptr;
+
+    VerifyOrExit(commandObj != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+
+    SuccessOrExit(err = commandObj->PrepareCommand(cmdParams));
+    writer = commandObj->GetCommandDataElementTLVWriter();
+    if (certChainType == 1)
+    {
+        SuccessOrExit(err = writer->Put(TLV::ContextTag(0), ByteSpan(dac_certificate, sizeof(dac_certificate))));
+    }
+    else if (certChainType == 2)
+    {
+        SuccessOrExit(err = writer->Put(TLV::ContextTag(0), ByteSpan(pai_certificate, sizeof(pai_certificate))));
+    }
+    else
+    {
+        SuccessOrExit(err = CHIP_ERROR_INTERNAL);
+    }
+    SuccessOrExit(err = commandObj->FinishCommand());
+
+exit:
+    if (status == EMBER_ZCL_STATUS_FAILURE)
+    {
+        emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: Failed CertChainRequest.");
+        emberAfSendImmediateDefaultResponse(status);
+    }
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Zcl, "Failed to encode response command: %s", ErrorStr(err));
+    }
+
+    return true;
+}
+
+bool emberAfOperationalCredentialsClusterAttestationRequestCallback(chip::app::Command * commandObj,
+                                                                    chip::ByteSpan attestationNonce)
+{
+    EmberAfStatus status    = EMBER_ZCL_STATUS_SUCCESS;
+    CHIP_ERROR err          = CHIP_NO_ERROR;
+    TLV::TLVWriter * writer = nullptr;
+    System::PacketBufferTLVWriter tlvWriter;
+    TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+    System::PacketBufferHandle certDeclaration;
+    System::PacketBufferHandle certDeclarationTbs;
+    System::PacketBufferHandle firmwareInfo;
+    System::PacketBufferHandle attestationElements;
+    System::PacketBufferHandle attestationElementsTbs;
+    Crypto::P256ECDSASignature signature;
+    Crypto::P256Keypair keypair;
+    Crypto::P256SerializedKeypair serializedKeypair;
+    PeerConnectionState * state;
+    ByteSpan attestationChallenge;
+    uint16_t certificationElement = 0;
+
+    emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: commissioner has requested Attestation");
+
+    app::CommandPathParams cmdParams = { emberAfCurrentEndpoint(), /* group id */ 0, ZCL_OPERATIONAL_CREDENTIALS_CLUSTER_ID,
+                                         ZCL_ATTESTATION_RESPONSE_COMMAND_ID, (chip::app::CommandPathFlags::kEndpointIdValid) };
+
+    // Fetch current admin
+    AdminPairingInfo * admin = retrieveCurrentAdmin();
+    VerifyOrExit(admin != nullptr, status = EMBER_ZCL_STATUS_FAILURE);
+
+    SuccessOrExit(err = serializedKeypair.SetLength(sizeof(dac_private_key) + sizeof(dac_public_key)));
+    memcpy(serializedKeypair, dac_public_key, sizeof(dac_public_key));
+    memcpy(serializedKeypair + sizeof(dac_public_key), dac_private_key, sizeof(dac_private_key));
+    SuccessOrExit(err = keypair.Deserialize(serializedKeypair));
+
+    certDeclarationTbs = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
```

Comment:
`System::PacketBuffer::kMaxSize` is too large here. The size of the CD is known in advance, so the allocation should be limited to its exact size.

Reply:
CD construction was removed - Updated.